### PR TITLE
Relax ingress await restrictions

### PR DESCRIPTION
**Pull request by viveklak**
### Proposed changes
Currently an ingress object is marked available when the backend endpoints are available and when the loadbalancer IP or host is available. The change here relaxes some assumptions made previously around ingress objects only referring to service backends or paths referencing endpoints that would emit events. The loadbalancer still a strict requirement but endpoint checks have been relaxed.

While I think this is the better default behavior, I wonder if we should add an annotation to opt-in to the previous stricter await logic?

### Related issues (optional)
Fixes https://github.com/pulumi/pulumi-kubernetes/issues/1810

### CODE REVIEW

File: provider/pkg/provider/helm_release.go [hunk 1]
```diff
@@ -212,7 +212,7 @@ func newHelmReleaseProvider(
 }
 
 func debug(format string, a ...interface{}) {
-	logger.V(3).Infof("[DEBUG] %s", fmt.Sprintf(format, a...))
+	logger.V(6).Infof("[DEBUG] %s", fmt.Sprintf(format, a...))
```
Comment (viveklak) [inline provider/pkg/provider/helm_release.go:215]: Sneaking this in. Happy to move to a separate PR if necessary.

File: provider/pkg/await/ingress.go [hunk 2]
```diff
@@ -336,21 +346,27 @@ func (iia *ingressInitAwaiter) checkIfEndpointsReady() (string, bool) {
 		}
 
 		for _, rule := range obj.Spec.Rules {
-			if rule.HTTP == nil {
-				iia.config.logStatus(diag.Error, fmt.Sprintf("expected value %q is unset for ingress: %s",
-					".spec.rules[*].http", obj.Name))
-				return apiVersion, false
+			var httpRules []networkingv1beta1.HTTPIngressPath
```
Comment (lblackstone) [inline provider/pkg/await/ingress.go:349]: Are we unmarshalling to the `v1beta1` struct? I would have expected to use the `v1` struct.

File: provider/pkg/await/ingress.go [hunk 3]
```diff
@@ -336,21 +346,27 @@ func (iia *ingressInitAwaiter) checkIfEndpointsReady() (string, bool) {
...
+
+			if rule.HTTP != nil {
+				httpRules = rule.HTTP.Paths
```
Comment (lblackstone) [inline provider/pkg/await/ingress.go:352]: `httpPaths` seems more accurate

File: provider/pkg/await/ingress.go [hunk 4]
```diff
@@ -443,15 +468,8 @@ func (iia *ingressInitAwaiter) processEndpointEvent(event watch.Event, settledCh
...
-		field := ".spec.rules[].http.paths[].backend.serviceName"
-		switch apiVersion {
-		case "networking.k8s.io/v1":
-			field = ".spec.rules[].http.paths[].backend.service.name"
-		}
-		messages = append(messages, fmt.Sprintf(
-			"Ingress has at least one rule that does not target any Service. "+
-				"Field '%v' may not match any active Service", field))
+	if _, ready := iia.checkIfEndpointsReady(); !ready {
+		messages = append(messages, "Ingress has at least one rule who's target endpoint didn't become available in time.")
```
Comment (lblackstone) [inline provider/pkg/await/ingress.go:472]:
```suggestion
		messages = append(messages, "Ingress has at least one rule with an unavailable target endpoint.")
```

File: provider/pkg/await/ingress_test.go [hunk 5]
```diff
@@ -58,44 +58,74 @@ func Test_Extensions_Ingress(t *testing.T) {
...
+			do: func(ingresses, services, endpoints chan watch.Event, settled chan struct{}, settlementGracePeriod, timeout chan time.Time) {
 				// Trigger timeout.
 				timeout <- time.Now()
 			},
 			expectedError: &timeoutError{
 				object: ingressInput("default", "foo", "foo-4setj4y6"),
 				subErrors: []string{
-					"Ingress has at least one rule that does not target any Service. " +
-						"Field '.spec.rules[].http.paths[].backend.serviceName' may not match any active Service",
+					"Ingress has at least one rule who's target endpoint didn't become available in time.",
```
Comment (lblackstone) [inline provider/pkg/await/ingress_test.go:97]:
```suggestion
					"Ingress has at least one rule with an unavailable target endpoint.",
```

**Review by lblackstone — CHANGES_REQUESTED** [review]

File: provider/pkg/await/ingress.go (hunk 2 above)
Comment (viveklak) [reply provider/pkg/await/ingress.go:349]: I guess it would depend on the preferred version in the kubernetes version? I do agree with you but lets keep this for now and reconsider at the next major release.

File: provider/pkg/await/ingress.go (hunk 3 above)
Comment (viveklak) [reply provider/pkg/await/ingress.go]: Fixed.

File: provider/pkg/await/ingress.go (hunk 4 above)
Comment (viveklak) [reply provider/pkg/await/ingress.go]: Fixed

File: provider/pkg/await/ingress_test.go (hunk 5 above)
Comment (viveklak) [reply provider/pkg/await/ingress_test.go]: Fixed

**Comment by viveklak** [comment]
@lblackstone Thoughts on the following?

> While I think this is the better default behavior, I wonder if we should add an annotation to opt-in to the previous stricter await logic?

**Comment by lblackstone** [comment]
> @lblackstone Thoughts on the following?
> 
> > While I think this is the better default behavior, I wonder if we should add an annotation to opt-in to the previous stricter await logic?

I think it's ok to loosen the behavior unconditionally for now, and we can reevaluate if we get user feedback about it. The result of an error here is most likely just a transient update failure if something downstream gets updated too early.

**Review by lblackstone — APPROVED** [review]
